components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/remote/RemoteConnectionSupport.java

```diff
@@ -0,0 +1,92 @@
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.proxy.remote;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+
+import org.sonatype.nexus.repository.httpclient.HttpClientFacet;
+import org.sonatype.nexus.repository.proxy.ProxyFacetSupport;
+import org.sonatype.nexus.repository.view.Payload;
+import org.sonatype.nexus.repository.view.payloads.BytesPayload;
+import org.sonatype.nexus.repository.view.payloads.HttpEntityPayload;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.io.ByteStreams;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.HttpStatus;
+import org.apache.http.StatusLine;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.conn.ConnectionPoolTimeoutException;
+import org.apache.http.util.EntityUtils;
+
+/**
+ * @since 3.0
+ */
+public abstract class RemoteConnectionSupport
+    extends ComponentSupport
+    implements RemoteConnection
+{
+  protected final ProxyFacetSupport proxyFacet;
+
+  public RemoteConnectionSupport(final ProxyFacetSupport proxyFacet) {
+    this.proxyFacet = proxyFacet;
+  }
+
+  public GetResult get(final URI uri) throws IOException {
+    HttpClient client = proxyFacet.getRepository().facet(HttpClientFacet.class).getHttpClient();
+
+    HttpGet request = new HttpGet(uri);
+    log.debug("Fetching: {}", request);
+
+    HttpResponse response = client.execute(request);
+    log.debug("Response: {}", response);
+
+    StatusLine status = response.getStatusLine();
+    log.debug("Status: {}", status);
+
+    Payload payload = null;
+    if (status.getStatusCode() == HttpStatus.SC_OK) {
+      HttpEntity entity = response.getEntity();
+      try {
+        log.debug("Entity: {}", entity);
+        final HttpEntityPayload httpEntityPayload = new HttpEntityPayload(response, entity);
+
+        payload = readFully(httpEntityPayload);
+      }
+      finally {
+        EntityUtils.consume(entity);
+      }
+    }
+    return new GetResult(status.getStatusCode(), payload);
+  }
+
+  protected boolean isRemoteUnavailable(final Exception e) {
+    if (e instanceof ConnectionPoolTimeoutException) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Read an incoming Payload into memory.
+   */
+  private Payload readFully(final Payload payload) throws IOException {
+    try (InputStream stream = payload.openInputStream()) {
+      return new BytesPayload(ByteStreams.toByteArray(stream), payload.getContentType());
+    }
+  }
```

Comment:
Alin inherited this from me; IIRC the re-readability of the payload was giving me a headache and I clearly missed a TODO to revisit this and clean it out.